consensus/src/chained_bft/chained_bft_smr.rs

```diff
@@ -88,61 +90,98 @@ impl<T: Payload> ChainedBftSMR<T> {
     fn start_event_processing(
         executor: Handle,
         mut epoch_manager: EpochManager<T>,
-        mut event_processor: EventProcessor<T>,
+        mut event_processor: Box<EventProcessor<T>>,
         mut pacemaker_timeout_sender_rx: channel::Receiver<Round>,
         network_task: NetworkTask<T>,
-        mut network_receivers: NetworkReceivers<T>,
+        all_events: Box<dyn Stream<Item = anyhow::Result<Event<ConsensusMsg>>> + Send + Unpin>,
+        data_request_sender: mpsc::UnboundedSender<ConsensusDataRequest>,
+        mut data_receiver: mpsc::UnboundedReceiver<(AccountAddress, FromNetworkMsg<T>)>,
     ) {
         let fut = async move {
             event_processor.start().await;
+
+            // Ask for new data on start-up.
+            let _ = data_request_sender.unbounded_send(ConsensusDataRequest::NewData);
+
             loop {
                 let pre_select_instant = Instant::now();
                 let idle_duration;
+
                 select! {
-                    proposal_msg = network_receivers.proposals.select_next_some() => {
-                        idle_duration = pre_select_instant.elapsed();
-                        event_processor.process_proposal_msg(proposal_msg).await;
-                    }
-                    block_retrieval = network_receivers.block_retrieval.select_next_some() => {
-                        idle_duration = pre_select_instant.elapsed();
-                        event_processor.process_block_retrieval(block_retrieval).await;
-                    }
-                    vote_msg = network_receivers.votes.select_next_some() => {
+                    (peer_id, consensus_msg) = data_receiver.select_next_some() => {
+
                         idle_duration = pre_select_instant.elapsed();
-                        event_processor.process_vote(vote_msg).await;
-                    }
+
+                        let current_epoch_msg = match epoch_manager.check_epoch(&peer_id, consensus_msg).await {
+                            Ok(EpochCheck::Current(msg)) => {
+                                // Since the epoch did not change,
+                                // and further processing of this EpochMsg by the event-processor
+                                // cannot influence this,
+                                // we can optimistically request
+                                // another message from network.
+                                let _ = data_request_sender.unbounded_send(ConsensusDataRequest::NewData);
+                                msg
+                            },
+                            Ok(EpochCheck::NotCurrent) => {
+                                // No further processing required of an EpochMsg,
+                                // so we just request more data and go back to the select above.
+                                let _ = data_request_sender.unbounded_send(ConsensusDataRequest::NewData);
```

Comment:
It isn't intuitive to me that we are asking the network for data.

Reply:
Why not?